M2/Macaulay2/m2/lists.m2

```diff
@@ -247,19 +247,10 @@ randomSubset(ZZ, ZZ) := (N, n) -> (
 	    while (N - t) * rawRandomRRUniform defaultPrecision >= n - m
 	    do t += 1;
 	    first (t, t += 1))))
--- when size not given, use binomial distribution first to determine size
 randomSubset ZZ := N -> (
-    if N < 0 then error "expected a nonnegative integer"
-    else if N < 30 then (
-	-- for small N, sum up N Bernoulli variates
-	n := 0;
-	scan(N, i -> n += random 2))
-    else (
-	-- for large N, use normal approximation
-	n = round(n/2 + sqrt(n/4)*rawRandomRRNormal defaultPrecision);
-	-- in the vanishingly unlikely case we end up outside [0, N]:
-	n = min(N, max(0, n)));
-    randomSubset(N, n))
+    if N < 0 then error "expected a nonnegative integer";
+    r := random 2^N;
+    for i to N - 1 list if r & (k := 2^i) == k then i else continue)
```

Comment:
if you really want to optimise you probably want something like
```
if first(r&1==0,r//=2) then ...
```

Reply:
though it's probably pointless in M2...

Reply:
I tried it out and it's actually an itty bit slower than the current implementation.  Maybe GMP optimizes computing powers of 2 or something.

Reply:
okay at least remove the needless assignment of `k`

Reply:
Agreed, `r & 2^i != 0` should suffice.

Reply:
Ok -- I'll switch it to that.  I won't bother running the tests again.  The one failing build was an unrelated ThreadedGB example